File: alchemist/alchemist-maps/src/main/java/it/unibo/alchemist/model/implementations/routes/GraphHopperRoute.java
/*
 * Copyright (C) 2010-2019, Danilo Pianini and contributors listed in the main project's alchemist/build.gradle file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.implementations.routes;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.graphhopper.GHResponse;
import com.graphhopper.PathWrapper;
import com.graphhopper.util.PointList;

import it.unibo.alchemist.model.implementations.positions.LatLongPosition;
import it.unibo.alchemist.model.interfaces.GeoPosition;
import it.unibo.alchemist.model.interfaces.TimedRoute;

/**
 */
public final class GraphHopperRoute implements TimedRoute<GeoPosition> {

    private static final long serialVersionUID = -1455332156736222268L;
    private final int numPoints;
    private final double distance, time;
    private final List<GeoPosition> points;

    /**
     * @param response
     *            the response to use
     */
    public GraphHopperRoute(final GHResponse response) {
        final List<Throwable> errs = response.getErrors();
        if (errs.isEmpty()) {
            final PathWrapper resp = response.getBest();
            time = resp.getTime() / 1000d;
            distance = resp.getDistance();
            final PointList pts = resp.getPoints();
            numPoints = pts.getSize();
            final List<GeoPosition> temp = new ArrayList<>(numPoints);
            for (int i = 0; i < pts.getSize(); i++) {
                temp.add(new LatLongPosition(pts.getLatitude(i), pts.getLongitude(i)));
            }
            points = Collections.unmodifiableList(temp);
        } else {
            final String msg = errs.stream().map(Throwable::getMessage).collect(Collectors.joining("\n"));
            throw new IllegalArgumentException(msg, errs.get(0));
        }
    }

    @Override
    public double length() {
        return distance;
    }

    @Override
    public GeoPosition getPoint(final int step) {
        return points.get(step);
    }

    @Override
    public List<GeoPosition> getPoints() {
        return points;
    }

    @Override
    public double getTripTime() {
        return time;
    }

    @Override
    public Iterator<GeoPosition> iterator() {
        return points.iterator();
    }

    @Override
    public Stream<GeoPosition> stream() {
        return points.stream();
    }

    @Override
    public int size() {
        return numPoints;
    }

}

File: alchemist-maps/src/main/kotlin/it/unibo/alchemist/model/maps/routes/GraphHopperRoute.kt
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.maps.routes

import com.google.common.collect.ImmutableList
import com.graphhopper.GHResponse
import com.graphhopper.util.shapes.GHPoint3D
import it.unibo.alchemist.model.GeoPosition
import it.unibo.alchemist.model.maps.TimedRoute
import it.unibo.alchemist.model.maps.positions.LatLongPosition
import java.util.concurrent.TimeUnit

/**
 * Models a route on a map, built upon the information provided by a query to GraphHopper.
 */
class GraphHopperRoute(
    from: GeoPosition,
    to: GeoPosition,
    response: GHResponse,
) : TimedRoute<GeoPosition> {
    private val distance: Double
    private val time: Double
    override val points: ImmutableList<GeoPosition>

    init {
        val errors = response.errors
        if (errors.isEmpty()) {
            val bestResponse = response.best
            val points = bestResponse.points.map { it.asPosition() }
            val initDistance = from.distanceTo(points.first())
            var pointSequence: Sequence<GeoPosition> = points.asSequence()
            var actualDistance = bestResponse.distance
            if (initDistance > 0) {
                actualDistance += initDistance
                pointSequence = sequenceOf(from) + pointSequence
            }
            val endingDistance = points.last().distanceTo(to)
            if (endingDistance > 0) {
                actualDistance += endingDistance
                pointSequence += sequenceOf(to)
            }
            // m / s, times are returned in ms
            val averageSpeed = bestResponse.distance * TimeUnit.SECONDS.toMillis(1) / bestResponse.time
            time = actualDistance / averageSpeed
            distance = actualDistance
            val builder = ImmutableList.builder<GeoPosition>()
            pointSequence.forEach(builder::add)
            this.points = builder.build()
        } else {
            val firstError = errors.first()
            val exception =
                IllegalArgumentException(
                    "Failure in the GraphHopper routing system when navigating from $from to $to, " +
                        "received response:\n$response",
                    firstError,
                )
            errors.asSequence().drop(1).forEach { exception.addSuppressed(it) }
            throw exception
        }
    }

    override fun length(): Double = distance

    override fun getPoint(step: Int): GeoPosition = points[step]

    override fun getTripTime(): Double = time

    override fun iterator(): MutableIterator<GeoPosition> = points.iterator()

    override fun size(): Int = points.size

    private companion object {
        private const val serialVersionUID = 0L

        private fun GHPoint3D.asPosition() = LatLongPosition(lat, lon)
    }
}
